File: 334-increasing-triplet-subsequence/334-increasing-triplet-subsequence.java
class Solution {
    public int lower_bound(ArrayList<Integer> lis,int num){
        if(lis.size() == 0) return 0;
        int ans = 0;
        int low = 0,high = lis.size() - 1;
        while(low <= high){
            int mid = low + (high - low)/2;
            if(lis.get(mid) > num){
                high = mid - 1;
            }
            else {
                ans = mid;
                low = mid + 1;
            }
        }
        
        if(ans < lis.size() && lis.get(ans) < num){
            ans++;
        }
        return ans;
    }
    public boolean increasingTriplet(int[] nums) {
        
        ArrayList<Integer> lis = new ArrayList<Integer>();
        
        for(int e:nums){
            int index = lower_bound(lis,e);
            if(index == lis.size()){
                lis.add(e);
            }
            else {
                lis.set(index,e);
            }
            if(lis.size() == 3) return true;
        }
        return false;
    }
}

File: 334-increasing-triplet-subsequence/334-increasing-triplet-subsequence.cpp
class Solution {
public:
    bool increasingTriplet(vector<int>& nums) {
        int n = nums.size();
        int first = INT_MAX,second = first;
        
        for(int i=0;i<n;i++){
            if(nums[i] <= first){
                first = nums[i];
            }
            else if(nums[i] <= second){
                second = nums[i];
            }
            else {
                return true;
            }
        }
        return false;
    }
};